### PR TITLE
Cola Dimensional Anomaly

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
Nanotrasen. while experimenting with bluespace tech, has located a dimension filled with..interesting..reagents. now this cola will leak into our dimension, giving players a fun risk reward with choosing to drink the soda

## Why It's Good For The Game
More chaos added to the rounds, able to possibly obtain rarer reagents for more creative traitor kills, uses the "fun" reagents category for something that people will actually see

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

cola_ground

cola_inspect

cola_ono

</details>

## Changelog
:cl:
add: Added the Space Cola INFINITE
add: Adds associated event
imageadd: new icon for the cola inf
/:cl:

### CODE REVIEW

**Comment by DrDuckedGoose** [comment]
It's like a strange pill generator?

**Comment by Fronsis** [comment]
Okay from my understanding this is a new random event in which.. around 5-20 cans of space-cola  INFINITE with random reagents and units spawns somewhere? I mean, sounds good enough, might as well give us doctors main a reason to use more calomel/pentetic/blood filter

**Comment by EvilDragonfiend** [comment]
fun type should be handled very carefully. I suggest using general type, and it gets fun type at a rare chance - like 5%?

**Comment by EvilDragonfiend** [comment]
also, you might want to encrypt the cola contents. consider referring to the hypospray code.

File: code/modules/events/anomaly_cola.dm [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/Initialize(mapload)
+	var/reagents = volume
+	while(reagents)
+		var/newreagent = rand(1, min(reagents, 30))
+		list_reagents += list(get_random_reagent_id(CHEMICAL_RNG_FUN) = newreagent)
+		reagents -= newreagent
+	. = ..()
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/open_soda(mob/user)  // different pop message copy-pasted
+	to_chat(user, "As you pull the tab off \the [src], an indescribable smell fill the air.") //warning
```
Comment (BrumalLake) [inline code/modules/events/anomaly_cola.dm]: fills

**Comment by etherware-novice** [comment]
@EvilDragonfiend wheres the code for that? i cant find

**Comment by etherware-novice** [comment]
never mind i may be dumb

**Comment by EvilDragonfiend** [comment]
I think you can still examine what's inside. try to check `/obj/item/reagent_containers/pill/floorpill`

**Comment by EvilDragonfiend** [comment]
or maybe it might be fine if you can only drink, not able to pour it into somewhere.

**Comment by EvilDragonfiend** [comment]
I think you need to put 5u~10u of cola at least, since tasting no cola doesn't make sense.

File: code/modules/events/anomaly_cola.dm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	prevent_grinding = TRUE
+	reagent_flags = AMOUNT_VISIBLE | REFILLABLE  // disables dumping chems into container, you can still fill with dangerous chems for a cool trick
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/Initialize(mapload)
+	var/reagents = volume - COLA_AMT
+	while(reagents)
+		var/newreagent = rand(1, min(reagents, 30))
+		var/category = CHEMICAL_RNG_GENERAL
+		if(prob(10))
+			category = CHEMICAL_RNG_FUN
```
Comment (DrDuckedGoose) [inline code/modules/events/anomaly_cola.dm]: Not heartbreaking to leave it the way it is but, this helps condense code.
*Someone* may consider it hard to read, however.
```suggestion
		var/category = prob(10) ? CHEMICAL_RNG_FUN : CHEMICAL_RNG_GENERAL
```

Comment (etherware-novice) [reply code/modules/events/anomaly_cola.dm]: alright thanks!

**Comment by etherware-novice** [comment]
what failed?

File: code/modules/events/anomaly_cola.dm [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#define COLA_AMT 5
+
+/datum/round_event_control/infcola
+	name = "Cola Infinite"
+	typepath = /datum/round_event/infcola
+	min_players = 1
+	max_occurrences = 10
+	var/atom/special_target
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]: This isn't used anywhere

File: code/modules/events/anomaly_cola.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	can_malf_fake_alert = TRUE
+
+/datum/round_event/infcola
+	announceWhen = 2
+
+/datum/round_event/infcola/announce(fake)
+	priority_announce("Our long-range anomaly scanners have detected leakage from a soda filled dimension. Nanotrasen is [fake ? "very" : "not"] responsible for any damages caused by these anomalous canisters.", "General Alert", SSstation.announcer.get_rand_alert_sound())
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]: Don't change it if its fake

File: code/modules/events/anomaly_cola.dm [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	if(reagents && reagents.reagent_list.len)
+		. += "<span class='notice'>You feel an urge to finish off the drink.</span>"
+
+
+/datum/round_event/infcola/start()
+	for(var/i in 1 to rand(5, 20))  // generates between 5-20 cans
+		var/obj/item/reagent_containers/food/drinks/soda_cans/inf/newCan = new(src)
+		var/turf/warp = find_safe_turf(extended_safety_checks = TRUE, dense_atoms = FALSE)
+		newCan.forceMove(warp)
+		do_smoke(location=warp)
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]: Group this with the typepath definition

File: code/modules/events/anomaly_cola.dm [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/open_soda(mob/user)  // different pop message copy-pasted
+	to_chat(user, "As you pull the tab off \the [src], an indescribable smell fills the air.") //warning
+	ENABLE_BITFIELD(reagents.flags, OPENCONTAINER)
+	playsound(src, "can_open", 50, 1)
+	spillable = TRUE
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/examine()
+	. = ..()
+	if(reagents && reagents.reagent_list.len)
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]:
```suggestion
	if(length(reagents?.reagent_list))
```

File: code/modules/events/anomaly_cola.dm [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+/obj/item/reagent_containers/food/drinks/soda_cans/inf
+	name = "Space Cola INFINITE"
+	desc = "Cola. Probably not from space. Proceed with caution. A no-tamper filter prevents the drink from being drained and resold."
+	icon_state = "inf"
+	list_reagents = list(/datum/reagent/consumable/space_cola = COLA_AMT)
+	prevent_grinding = TRUE
+	reagent_flags = AMOUNT_VISIBLE | REFILLABLE  // disables dumping chems into container, you can still fill with dangerous chems for a cool trick
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/Initialize(mapload)
+	var/reagents = volume - COLA_AMT
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]: isn't reagents already a var on the object? Could get confusing if you override the variable locally when reagents already exists somewhere else

Comment (PowerfulBacon) [reply code/modules/events/anomaly_cola.dm]: reagents_left would be more suitable.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/modules/events/anomaly_cola.dm (hunk 5 above)
Comment (etherware-novice) [reply code/modules/events/anomaly_cola.dm]: hi, where is the typepath definition?

Comment (PowerfulBacon) [reply code/modules/events/anomaly_cola.dm]: Where it says and the variables are defined
/datum/typepath

File: code/modules/events/anomaly_cola.dm [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	typepath = /datum/round_event/infcola
+	min_players = 1
+	max_occurrences = 10
+	can_malf_fake_alert = TRUE
+
+/datum/round_event/infcola
+	announceWhen = 2
+
+/datum/round_event/infcola/announce(fake)
+	priority_announce("Our long-range anomaly scanners have detected leakage from a soda filled dimension. Nanotrasen is not responsible for any damages caused by these anomalous canisters.", "General Alert", SSstation.announcer.get_rand_alert_sound())
```
Comment (PowerfulBacon) [inline code/modules/events/anomaly_cola.dm]: A soda filled dimension? This needs to me made to align with lore / not be memey.

Comment (etherware-novice) [reply code/modules/events/anomaly_cola.dm]: ok what about space cola co experimenting with bluespace tech and causing this

**Comment by PowerfulBacon** [comment]
Yea, them attempting to teleport cola to another factory to save on shuttle costs and accidentally sending it to the station sounds more reasonable.

**Comment by etherware-novice** [comment]
alrighty coolo

File: code/modules/events/anomaly_cola.dm [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+/datum/round_event/infcola/start()
+	for(var/i in 1 to rand(5, 20))  // generates between 5-20 cans
+		var/obj/item/reagent_containers/food/drinks/soda_cans/inf/newCan = new(src)
+		var/turf/warp = find_safe_turf(extended_safety_checks = TRUE, dense_atoms = FALSE)
+		newCan.forceMove(warp)
+		do_smoke(location=warp)
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf
+	name = "Space Cola INFINITE"
+	desc = "Cola. Probably not from space. Proceed with caution. A no-tamper filter prevents the drink from being drained and resold."
```
Comment (itsmeow) [inline code/modules/events/anomaly_cola.dm]: tamper-proof container

File: code/modules/events/anomaly_cola.dm [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	while(reagents_left)
+		var/newreagent = rand(1, min(reagents_left, 30))
+		var/category = prob(10) ? CHEMICAL_RNG_FUN : CHEMICAL_RNG_GENERAL
+
+		list_reagents += list(get_random_reagent_id(category) = newreagent)
+		reagents_left -= newreagent
+	. = ..()
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/open_soda(mob/user)  // different pop message copy-pasted
+	to_chat(user, "As you pull the tab off \the [src], an indescribable smell fills the air.") //warning
```
Comment (itsmeow) [inline code/modules/events/anomaly_cola.dm]: `open \the [src]` or `pull off the tab`

File: code/modules/events/anomaly_cola.dm [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+	ENABLE_BITFIELD(reagents.flags, OPENCONTAINER)
+	playsound(src, "can_open", 50, 1)
+	spillable = TRUE
+
+/obj/item/reagent_containers/food/drinks/soda_cans/inf/examine()
+	. = ..()
+	if(length(reagents?.reagent_list))
+		. += "<span class='notice'>You feel an urge to finish off the drink.</span>"
```
Comment (itsmeow) [inline code/modules/events/anomaly_cola.dm]: remove `off`

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by PowerfulBacon** [comment]
Unresolved reviews

**Comment by PowerfulBacon** [comment]
I'm not a fan of this as I don't think it makes sense in the universe. I'd like to limit the amount of large scale silly things as it really kills the atmosphere and creates an environment where its harder to RP, since the world is less believable.

**Comment by etherware-novice** [comment]
come on!! please

**Comment by Penwin0** [comment]
Too memey, something we are attempting to slowly move away from, event itself not too interesting essentially just spawning a bunch of floorpill equivalents. 

I thought I had already added a key to this.

**Pull request closed without merging**